### PR TITLE
Several improvements to SV contig alignment configuration picker

### PR DESCRIPTION
**Pull request by SHuang-Broad**
Fix a particular "bug" observed that caused us dozens of variants:

When an assembly contig has a unique region covered by one or more MQ 0 alignments, the current configuration picker classifies such contig as "ambiguous". 
This PR fix this problem by implementing a tie breaker saying that for such contigs, we prefer the unique configuration with no MQ 0 alignments; that is, this tie breaker only saves contigs have one unique configuration with all non-MQ-0 mappings.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryUtils.java [hunk 1]
```diff
@@ -78,4 +91,26 @@ private static void evaluateIntervalsAgainstTruth(final List<SVInterval> assembl
...
+        if (nonCanonicalContigNamesFile!= null) {
+
+            try (final Stream<String> nonCanonical = Files.lines(IOUtils.getPath((nonCanonicalContigNamesFile)))) {
+                return new HashSet<>( Sets.difference(dictionary.getSequences().stream().map(SAMSequenceRecord::getSequenceName).collect(Collectors.toSet()),
+                        nonCanonical.collect(Collectors.toSet())) );
+            } catch ( final IOException ioe ) {
+                throw new UserException("Can't read nonCanonicalContigNamesFile file "+nonCanonicalContigNamesFile, ioe);
+            }
+        } else {
+            final List<String> first22ChromosomesNum = IntStream.range(0, 23).mapToObj(String::valueOf).collect(Collectors.toList());
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryUtils.java]: Can we not do this? It's better to require that either you have the file or you don't. If you don't have the file just treat everything as canonical. I don't want to hard code any chromosome names in the system, and I definitely don't want the tool to have any opinion on whether hg19/b37 chromosome names have "chr" in them. I know the code has been there for a while but somehow I don't remember reviewing this; maybe I just missed it in the original large pull request.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 2]
```diff
@@ -48,8 +49,52 @@
...
+        return breakTieByPreferringLessAlignments(parsedContigs);
+    }
+
+    //==================================================================================================================
+
+    // TODO: 1/31/18 this is a proof of concept, we could expand on allowing a customizable MQ threshold, and save the bad mappings as String representations for later annotation
+    /**
+     * For contigs with more than 1 best-scored configurations as determined by
+     * {@link #pickBestConfigurations(AlignedContig, Set, Double)},
+     * save the contigs that has one and only one configuration that
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: "that have"

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 3]
```diff
@@ -48,8 +49,52 @@
...
+     * has all mapping quality above 0.
+     */
+    private static JavaRDD<AlignedContig>
+    breakTieByPreferringLessAlignments(final JavaRDD<AlignedContig> contigsWithAmbiguousPictures) {
+        final JavaRDD<Iterable<AlignedContig>> contigsGroupedByName =
+                contigsWithAmbiguousPictures.groupBy(contig -> contig.contigName).values();
+
+        final Tuple2<JavaRDD<Iterable<AlignedContig>>, JavaRDD<Iterable<AlignedContig>>> split =
+                RDDUtils.split(contigsGroupedByName, AssemblyContigAlignmentsConfigPicker::hasOneAndOnlyOneAllNonZeroMQs, false);
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: Do we need to split the rdd, process each half separately, and then union? That seems really inefficient. Can't you write this as a series of maps and filters?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 4]
```diff
@@ -48,8 +49,52 @@
...
+
+        final JavaRDD<AlignedContig> savedOnes =
+                split._1
+                        .map(representations ->
+                                Utils.stream(representations)
+                                        .filter(representation ->
+                                                representation.alignmentIntervals.stream().mapToInt(ai -> ai.mapQual).min().orElse(0) != 0)
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: This line of code appears in `hasOneAndOnlyOneAllNonZeroMQs` as well, maybe extract a method.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 5]
```diff
@@ -48,8 +49,52 @@
...
+                                        .findFirst()
+                                        .orElseThrow(() -> new GATKException(""))
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: Add message to exception.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 6]
```diff
@@ -48,8 +49,52 @@
...
+     */
+    private static JavaRDD<AlignedContig>
+    breakTieByPreferringLessAlignments(final JavaRDD<AlignedContig> contigsWithAmbiguousPictures) {
+        final JavaRDD<Iterable<AlignedContig>> contigsGroupedByName =
+                contigsWithAmbiguousPictures.groupBy(contig -> contig.contigName).values();
+
+        final Tuple2<JavaRDD<Iterable<AlignedContig>>, JavaRDD<Iterable<AlignedContig>>> split =
+                RDDUtils.split(contigsGroupedByName, AssemblyContigAlignmentsConfigPicker::hasOneAndOnlyOneAllNonZeroMQs, false);
+
+        final JavaRDD<AlignedContig> savedOnes =
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: Why are these called `savedOnes` and `unsavedOnes`? The names don't seem accurate to me (we end up keeping all of the `unsavedOnes`, no?

**Comment by cwhelan** [comment]
Is it guaranteed that one of the configurations won't include any of the MQ0 regions? Why is that?

**Comment by cwhelan** [comment]
When you say this "caused us dozens of variants" do you mean true positives or false negatives?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryUtils.java (hunk 1 above)
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryUtils.java]: removed

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java (hunk 2 above)
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java (hunk 3 above)
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: updated

**Comment by SHuang-Broad** [comment]
@cwhelan Thanks for the review! And I apologize for not clearly stating what problem is getting fixed here.

I've addressed the comments in separate commit, changed the implementation, made more improvements that were discovered while reviewing the variants 
(https://github.com/SHuang-Broad/GATK-SV-callset-regressionTest/tree/master/Evaluation/Analysis/masterVSfeature/notes.xlsx)
The implemented fixes are:
* for removing the hard-coded/explicit mentioning of "chr" in non-canonical versions, it is now fixed in 5eff782e4d582d516004fba2cee7535d984b1540
* for contigs whose alignments paint ambiguous picture, i.e. multiple alignment configurations offer equally good explanation:
	1. if only one configuration has all alignment with MQ above a specified threshold, it is favored; this is implemented in ecc31f5fbec4e524b401fc9474a3a1b7ab08c561
	2. if one configuration has alignment to non-canonical chromosome that explains the contig better than would-be-event-inducing mappings to canonical chromosomes, the canonical mappings are saved but the better non-canonical mappings are saved as SA tag as in SAM spec, and the VCF record produced is annotated accordingly; this is implemented in 65cdb523a2f9fa2026334713fed45381d76ffc82
* fixed a bug where sometimes an assembly contig as several alignments, only one of which has non-mediocre MQ but at the sametime this alignment contains a large gap, such contigs were previously incorrectly filtered away, they are now salvaged by commit b6b2f197b112981e00efd9d415f010c024d31b36

So, for the FN variants (FN in the sense that they are captured in the stable version of our interpretation tool but now goes missing in the experimental interpretation tool)
that were curated in the above-mentioned review, only the following ones are not salvaged, with plans or comments attached

```
asm012854:tig00000	missing	classified as "incomplete"; fixable by finishing the last TODO in AssemblyContigAlignmentSignatureClassifier (same problem as face by group represented by asm002398:tig00001)
asm014580:tig00018	missing	classified as "incomplete"; fixable by finishing the last TODO in AssemblyContigAlignmentSignatureClassifier (same problem as face by group represented by asm002398:tig00001)
asm008185:tig00000	missing	classified as "ambiguous"; unsure of how to deal with a general rule--exact same part of read coverred by one mapping to canonical, one to non-canonical (better AS); if save the canonical, a deletion would be called
asm018220:tig00004	missing	classified as "ambiguous"; salvagable if re-analysed through old interpretation tool
asm026229:tig00000	missing	classified as "ambiguous"; fixable if implement a special rule making all split alignments on non-canonical chromosomes as "bad"
```

They are relatively simple fixes so I clumped them into a single PR.
If you prefer to review them as separate PR's. I can do that too.
Thanks!

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigWithFineTunedAlignments.java [hunk 7]
```diff
@@ -29,20 +30,29 @@
...
-    public AssemblyContigWithFineTunedAlignments(final AlignedContig contig,
-                                                 final List<String> insertionMappings,
-                                                 final String saTAGForGoodMappingToNonCanonicalChromosome) {
-        this.contig = contig;
-        this.insertionMappings = Utils.nonNull(insertionMappings);
+    // {@code goodMappingToNonCanonicalChromosome} could be null
+    public AssemblyContigWithFineTunedAlignments(@Nonnull final AlignedContig contig,
+                                                 @Nonnull final List<String> insertionMappings,
+                                                 final AlignmentInterval goodMappingToNonCanonicalChromosome) {
+        this(Utils.nonNull(contig), Utils.nonNull(insertionMappings),
```
Comment (cwhelan) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigWithFineTunedAlignments.java]: Do you need both calls to `Utils.nonNull` and the annotations in the method signature?

Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigWithFineTunedAlignments.java]: maybe I'm just paranoid.... removed the annotation

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 8]
```diff
@@ -461,7 +509,42 @@ private static boolean gappedAlignmentOffersBetterCoverage(final AlignmentInterv
...
+     *  one configuration has split alignments to canonical chromosomes, hence indicating the SV, whereas
+     *  the other configuration has a single, often very good (or even better) alignment to a non-canonical chromosome.
+     * We send down the chimeric alignment configuration for inference, but notes down that a non-canonical chromosome
+     * in the reference bundle input could have already captured the SV on this sample.
+     *
+     * @return  {@code null} if the non-canonical chromosome mapping doesn't offer a better score,
+     *          otherwise the non-canonical chromosome mapping
+     */
+    @VisibleForTesting
+    static AlignmentInterval specialChanelForSingleNonCanonicalMappings(final Set<String> canonicalChromosomes,
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: changed as suggested

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 9]
```diff
@@ -170,28 +185,41 @@ public boolean equals(final Object o) {
...
      */
     @VisibleForTesting
     static List<GoodAndBadMappings> pickBestConfigurations(final AlignedContig alignedContig,
                                                            final Set<String> canonicalChromosomes,
                                                            final Double scoreDiffTolerance) {
+        // nothing to score if only one alignment
+        if (alignedContig.alignmentIntervals.size() == 1) {
+            return Collections.singletonList(
+                    new GoodAndBadMappings(Collections.singletonList(alignedContig.alignmentIntervals.get(0)),
+                                           Collections.emptyList(), null)
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: ah, right! Thanks for catch that!

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 10]
```diff
@@ -461,7 +509,42 @@ private static boolean gappedAlignmentOffersBetterCoverage(final AlignmentInterv
...
+     * In other words, the non-canonical chromosome captures an SV--relative to the canonical chromosomes--of relatively high frequency.
+     *
+     * The sample under analysis could have the allele of this non-canonical version,
+     * and be marked as having an SV on the corresponding location on the canonical chromosome.
+     *
+     * An assembly contig from this sample may have two equally well scored alignment configurations, where
+     *  one configuration has split alignments to canonical chromosomes, hence indicating the SV, whereas
+     *  the other configuration has a single, often very good (or even better) alignment to a non-canonical chromosome.
+     * We send down the chimeric alignment configuration for inference, but notes down that a non-canonical chromosome
+     * in the reference bundle input could have already captured the SV on this sample.
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: corrected. Thanks for the explanation!

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 11]
```diff
@@ -461,7 +509,42 @@ private static boolean gappedAlignmentOffersBetterCoverage(final AlignmentInterv
         }
     }
 
-    // TODO: 3/1/18 implement another simple fix that classify contigs with the following signature as ambiguous
-    // one configuration with single good mapping to non-canonical chromosome,
-    // another configuration with split alignments to canonical chromosomes
+    /**
+     * There are locations on the non-canonical chromosomes of the HG38 reference that is similar to a location
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: corrected

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 12]
```diff
@@ -24,8 +24,22 @@
...
 public class AssemblyContigAlignmentsConfigPicker {
 
-    public static final int ALIGNMENT_MAPQUAL_THREHOLD = 20;
-    public static final int ALIGNMENT_READSPAN_THRESHOLD = 10;
+    /**
+     * A filter that is used to remove contigs upfront which doesn't meet the following criteria
+     * either:
+     *  has only 1 mapping, with MQ strictly above this threshold
+     * or:
+     *  has more than 1 mappings, but only 1 mapping has MQ strictly above this threshold and it has a large gap in it.
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: corrected

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 13]
```diff
@@ -322,4 +368,30 @@ private static boolean gappedAlignmentOffersBetterCoverage(final AlignmentInterv
...
+     * {@link #pickBestConfigurations(AlignedContig, Set, Double)},
+     * save the contigs that has one and only one configuration that
+     * has all mapping quality strictly above the specified {@code threshold}.
+     * Example:
+     *  if a contig has two equal scored configurations with MQ's {10, 60, 60}, and {60, 60},
+     *  this function will favor/pick the {60, 60} configuration if the threshold is 10,
+     *  hence remove the ambiguity.
+     */
+    @VisibleForTesting
+    static List<GoodAndBadMappings> breakTieByPreferringLessAlignments(
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: Sorry for not squashing the later bug fix commit here.
You are right, it is a bad choice of name for the method. 
But I did change the name as suggested. Thanks!

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java [hunk 14]
```diff
@@ -116,20 +116,58 @@ private static boolean notDiscardForBadMQ(final AlignedContig contig) {
...
+                .map(nameSeqAndConfigurations -> new Tuple2<>(nameSeqAndConfigurations._1, breakTieByPreferringLessAlignments(nameSeqAndConfigurations._2, 0)))
+                .map(nameSeqAndConfigurations -> new Tuple2<>(nameSeqAndConfigurations._1, nameSeqAndConfigurations._2.stream().map(r -> r.goodMappings).collect(Collectors.toList()))) // todo temporary, later need to change interface of the next map
                 .flatMap(AssemblyContigAlignmentsConfigPicker::reConstructContigFromPickedConfiguration);
     }
 
+    /**
+     * After configuration scoring and picking, the original alignments can be classified as
+     * good and bad mappings:
+     * good: the ones present the picked configuration
+     * bad: the brings more noise than information,
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/alignment/AssemblyContigAlignmentsConfigPicker.java]: corrected

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryUtils.java [hunk 15]
```diff
@@ -85,4 +97,20 @@ private static void evaluateIntervalsAgainstTruth(final List<SVInterval> assembl
...
     }
+
+    public static Set<String> getCanonicalChromosomes(final String nonCanonicalContigNamesFile,
+                                                      @Nonnull final SAMSequenceDictionary dictionary) {
+        final Stream<String> allContigsStream = Utils.nonNull(dictionary).getSequences().stream().map(SAMSequenceRecord::getSequenceName);
+        if (nonCanonicalContigNamesFile == null)
+            return allContigsStream.collect(Collectors.toCollection(LinkedHashSet::new));
+
+        try (final Stream<String> nonCanonical = Files.lines(IOUtils.getPath(( Utils.nonNull(nonCanonicalContigNamesFile) )))) {
+            final HashSet<String> toIgnore = nonCanonical.collect(Collectors.toCollection(HashSet::new));
```
[the remark this answers is not in the document]
Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoveryUtils.java]: removed. Thanks!

**Pull request merged**